x/exchange/keeper/orders.go

```diff
@@ -0,0 +1,711 @@
+func (k Keeper) getBidOrders(store sdk.KVStore, marketID uint32, orderIDs []uint64, seller string) ([]*exchange.Order, error) {
+	var errs []error
+	orders := make([]*exchange.Order, 0, len(orderIDs))
+
+	for _, orderID := range orderIDs {
+		order, oerr := k.getOrderFromStore(store, orderID)
+		if oerr != nil {
+			errs = append(errs, oerr)
+			continue
+		}
+		if order == nil {
+			errs = append(errs, fmt.Errorf("order %d not found", orderID))
+			continue
+		}
+		if !order.IsBidOrder() {
+			errs = append(errs, fmt.Errorf("order %d is type %s: expected bid", orderID, order.GetOrderType()))
+			continue
+		}
+
+		bidOrder := order.GetBidOrder()
+		orderMarketID := bidOrder.MarketId
+		buyer := bidOrder.Buyer
+
+		if orderMarketID != marketID {
+			errs = append(errs, fmt.Errorf("order %d market id %d does not equal requested market id %d", orderID, orderMarketID, marketID))
+			continue
+		}
+		if buyer == seller {
+			errs = append(errs, fmt.Errorf("order %d has the same buyer %s as the requested seller", orderID, buyer))
+			continue
+		}
+
+		orders = append(orders, order)
+	}
+
+	return orders, errors.Join(errs...)
+}
+
+// placeHoldOnOrder places a hold on an order's funds in the owner's account.
+func (k Keeper) placeHoldOnOrder(ctx sdk.Context, order exchange.OrderI) error {
+	orderID := order.GetOrderID()
+	orderType := order.GetOrderType()
+	owner := order.GetOwner()
+	ownerAddr, err := sdk.AccAddressFromBech32(owner)
+	if err != nil {
+		return fmt.Errorf("invalid %s order %d owner %q: %w", orderType, orderID, owner, err)
+	}
+	toHold := order.GetHoldAmount()
+	err = k.holdKeeper.AddHold(ctx, ownerAddr, toHold, fmt.Sprintf("x/exchange: order %d", orderID))
+	if err != nil {
+		return fmt.Errorf("error placing hold for %s order %d: %w", orderType, orderID, err)
+	}
+	return nil
+}
+
+// releaseHoldOnOrder releases a hold that was placed on an order's funds in the owner's account.
+func (k Keeper) releaseHoldOnOrder(ctx sdk.Context, order exchange.OrderI) error {
+	orderID := order.GetOrderID()
+	orderType := order.GetOrderType()
+	owner := order.GetOwner()
+	ownerAddr, err := sdk.AccAddressFromBech32(owner)
+	if err != nil {
+		return fmt.Errorf("invalid %s order %d owner %q: %w", orderType, orderID, owner, err)
+	}
+	held := order.GetHoldAmount()
+	err = k.holdKeeper.ReleaseHold(ctx, ownerAddr, held)
+	if err != nil {
+		return fmt.Errorf("error releasing hold for %s order %d: %w", orderType, orderID, err)
+	}
+	return nil
+}
+
+// GetOrder gets an order. Returns nil, nil if the order does not exist.
+func (k Keeper) GetOrder(ctx sdk.Context, orderID uint64) (*exchange.Order, error) {
+	return k.getOrderFromStore(k.getStore(ctx), orderID)
+}
+
+// CreateAskOrder creates an ask order, collects the creation fee, and places all needed holds.
+func (k Keeper) CreateAskOrder(ctx sdk.Context, askOrder exchange.AskOrder, creationFee *sdk.Coin) (uint64, error) {
+	if err := askOrder.Validate(); err != nil {
```

Comment:
Does this need to be validated again? I think it's done in ValidateBasic.

Reply:
Yup. As a public keeper function, other modules/stuff might try to use this function, and we don't want anyone to be able to create an invalid order.